Add pt-galera-log-explainer

ylacancellera opened the pull request:
Pushing a new tool: pt-galera-log-explainer

**Goal of this PR**
The goal is discussing the state, if it's ready/stable enough for toolkits
<br/><br/>

**Main features:**

The goal is to aggregate and analyze Galera logs from various nodes and help analyzing them
- Listing important logs in chronological order, cluster-wide
- Filtering main events by themes
- Map node identification information between each others: node uuids, ips, node names
- Tracking replication failures
<br/><br/>

**How to read the code/how it works:**

Basically, it's a "grep" command on each files (in internal.go). 
"grep" arguments are compiled using anonymous structs, organized in maps by themes (regex directory, e.g regex/views.go).
Those structs have regexes, and associated handlers to build the file context and something to print eventually (see types/regex.go).
The choice of "grep" is detailed in internal.go

It iterates on grep results, and create a "timeline" (types/timeline.go), ordered by timestamp. A timeline is a slice of "loginfo", which is a struct everything there is to know for a log line (types/loginfo.go). it will store its copy of the file context (not a pointer, else some info like states or sst would be lost) and some metadata.

When all goes well, when finishing a log we can identify it by wsrep_node_name, and "timelines" will be put in a map identified by wsrep_node_name. Any other log with the same wsrep_node_name will be merged with the other one (types/timeline.go, MergeTimeline)

(For the main subcommand "list") once finished, we have a map of timelines. Each "timelines" key will be a column to print
To print (display/timelinecli.go), it will dequeue events. It will iterate on each "timeline", get the next minimum timestamp and that will be the next event to print. (actually eventS, in case 2 have the same timestamps, which happens on some version)
Once printed, the event is deleted from its timeline
Other columns will have a placeholder "|", that will be colored depending the current state (donor, synced, ko, ...)
<br/><br/>
**Quircks to notify when reading**

The message to print looks obfuscated, everything is done with closures
It's just to inject the latest context everytime. Basically, instead of printing "172.17.0.2 left" right away, it waits to have complete context so that it will be able to map "172.17.0.2" to something better like "node1".
(same logics applies to sst, every views changes, conflicts)
EDIT: those translations are stored and handled in "translate" package. It's basically maps associating something with a "value"/timestamp pair.

Regexes structures can look over-engineered (maybe they are?). See types/regex.go for the struct, regex/applicative.go to see an example of why it's structured this way regarding closures. It enables to keep business logic at the same place and avoid any exception
<br/><br/>
**Tests**

- every regexes are tested, see regex/regex_test.go. It tests if "grep" works, if context is updated correctly and output is correct
- there are no complete functional tests. could be done, with a bit of discussion
- by the nature of this tool, some logs can have unexpected results (rare logs, missing information, log only printing warnings). Some won't be identified correctly (mostly because of lack of info)


**note**
- not a single commit to check/improve performance was done. (a tiny bit with regexes, else it's quickly slow) I preferred keeping things simple before optimizing.
- will add small bugfixes here or simplifications as needed
- main.go readability is being improved

- [x] The contributed code is licensed under GPL v2.0
- [ ] Contributor Licence Agreement (CLA) is signed
- [ ] util/update-modules has been ran
- [ ] Documentation updated
- [x] Test suite update

ylacancellera committed:
Add pt-galera-log-explainer

ylacancellera committed:
Improve: "not found" error handling

ylacancellera committed:
Fix: forgot to inherit conflicts

ylacancellera committed:
Add: merge by directory, Simplified: timeline merges

ylacancellera committed:
Simplified the main loop

ylacancellera commented:
todo: 
* version cmd: should be done with --version, not "version". and it's broken when using the make command

Adi-78 reviewed (changes requested):
Please take any existing go tool and make this tool as version compatible to same as other go tool.

Adi-78 reviewed (changes requested):
Please add pt in usage which is missing.

svetasmirnova reviewed (changes requested):
README files are used to create user manuals for the tools. They are still markdown but with Percona extensions Please check standard headings and format in the existing tools, e.g. pt-mongodb-summary or pt-k8s-debug-collector, and adjust the README file. It is also better to rename it to README.rst

Adi-78 reviewed (changes requested):
panic: runtime error: index out of range [-1], seen after running below set of command:

./pt-galera-log-explainer list --all --since=2023-05-28T21:18:22.070376-05:00  --until=2023-05-29T08:19:02.849987-05:00 $HOME/*.log

Adi-78 reviewed (commented):
Repeating information is getting printed like (identifier, current path, last known ip, last known name, mysql version) in the end of log o/p which might be change.

Adi-78 reviewed (commented):
./pt-galera-log-explainer list --all --since=2023-05-28T21:18:22.070376-05:00  --until=2023-05-29T08:19:02.849987-05:00 $HOME/test_pt/*20230528.log -vvv so verbosity options like -v, -vv, -vvv should be improved

Adi-78 reviewed (changes requested):
$:~/percona-toolkit/bin$ ./pt-galera-log-explainer list --all --since=2023-05-28T21:18:22.070376-05:00  --until=2023-05-29T08:19:02.849987-05:00 $HOME/test_pt/*20230528.log --pxc-operator
{"level":"error","component":"extractor","path":"$HOME/test_pt/CS0036736_node0_mysqld.20230528.log","error":"Found nothing","time":1694523083,"message":"execGrepAndIterate returned error"}
{"level":"error","component":"extractor","path":"$HOME/test_pt/CS0036736_node1_mysqld.20230528.log","error":"Found nothing","time":1694523083,"message":"execGrepAndIterate returned error"}
{"level":"error","component":"extractor","path":"$HOME/test_pt/CS0036736_node2_mysqld.20230528.log","error":"Found nothing","time":1694523083,"message":"execGrepAndIterate returned error"}
panic: runtime error: index out of range [-1]

ylacancellera committed:
Fix: crash when timeline is empty

It was due to a silly regression when reformatting the main.go
The function iterating was doing too many things, and returning an error
when nothing was found, and a "continue" was done on the main
"timelineFromPaths" loop
It is now a simple foreach loop that does not return error so we have to
check if the localtimeline slice is empty

ylacancellera committed:
Fix: --version formatting

ylacancellera committed:
Fix: error msg with uppercase, usage missed pt-

ylacancellera commented:
> Please take any existing go tool and make this tool as version compatible to same as other go tool.

Done: https://github.com/percona/percona-toolkit/pull/669/commits/1ca6cb8d80a2d4bcf61392a45e475e1a9c539fed
```
pt-galera-log-explainer
Version 3.5.0
Build: 2023-09-19T10:16:30+0200 using go1.20.5
Commit: 1dd85793c25662ad4d5f117d8dad856221af7c87
```
(Version is just an example)

> panic: runtime error: index out of range [-1], seen after running below set of command:

Fixed in https://github.com/percona/percona-toolkit/pull/669/commits/063178d6462aafe82a202b906ccf6e89144d7d54
Silly regression introduced in the latest refactoring https://github.com/percona/percona-toolkit/pull/669/commits/1ca57cf76cc99b47e9b88dcbfd5fad9ae7576fe2
Working on adding good unit tests for this case (--since, --until, merging logs with some empty)

> Please add pt in usage which is missing.

Done

ylacancellera commented:
> Repeating information is getting printed like (identifier, current path, last known ip, last known name, mysql version) in the end of log o/p which might be change.

It is done with a simple 
```
if linecount >= 50 {
		fmt.Fprintln(w, separator(keys))
		fmt.Fprintln(w, headerNodes(keys))
		fmt.Fprintln(w, headerFilePath(keys, currentContext))
		fmt.Fprintln(w, headerIP(keys, currentContext))
		fmt.Fprintln(w, headerName(keys, currentContext))
		fmt.Fprintln(w, headerVersion(keys, currentContext))

}
```
Small consideration: most of the time it's the latest lines we read, because we read logs when something happened.
Keeping only the top header, even with a huge output, could force a huge scroll up that IMO could be avoided
That's my 2cnts, I would rather keep the bottom one only or increase "50".

If you both agree this is weird that way, I'll remove

ylacancellera commented:
> ./pt-galera-log-explainer list --all --since=2023-05-28T21:18:22.070376-05:00 --until=2023-05-29T08:19:02.849987-05:00 $HOME/test_pt/*20230528.log -vvv so verbosity options like -v, -vv, -vvv should be improved

Agreed. Let me explain, I'm interested for an input
**-v:** ON by default, it will show non-essential logs like "node is suspected to be down" , "sst request was accepted by X", "missing grastate.dat". So helpful things still, but we could do without
**-vv:** will show "useless" logs that only helped to build the names/ip/version context.
**-vvv:** straight debug mode for the script

So, maybe -v should be the "normal" mode and everything else shifted to the left ? Or just a single -v mode and that's it ?

ylacancellera committed:
Add: regression tests

ylacancellera committed:
Fix: "not found" should not be an error

if it shows nothing, it can be understood as easily as a grep returning
nothing
adding an error log is too verbose

ylacancellera committed:
Add: merge regression tests

ylacancellera commented:
Regression tests are here. They are complete enough for meaningful tests, they also show some current limitations which will be improved later (some missing information still) 

README still must be modified

ylacancellera committed:
Change: README to pt standard, regex-list --json

ylacancellera committed:
Add: --no-color regression test

ylacancellera committed:
Update README.rst

lalitvc commented:
It is good to have support for test environments with the same hostname and without wsrep node name.

ylacancellera committed:
Remove: extra space in --version out

ylacancellera committed:
Add: requirements in README.rst

ylacancellera committed:
Fix: don't cut node name if it's an ip

rare case, but it was not producing great results
regression tests will come for this

ylacancellera committed:
Add: proper error handling for files

ylacancellera committed:
Fix: pointer dereference if votes was missing

ylacancellera committed:
Change: simplify verbose mode

ylacancellera committed:
Add: concurrent SSTs handling

It is a thing: 2 nodes joining at the same time, with 2 JOINERs and 2
DONORs cluster-wide
It can happen on operators with 2 garbd joining at the same time

Before, pt-galera-log-explainer was using SST metadata naively.
Basically if a node was DONOR and we found a "transfer completed"
message, we assumed the donor name we found is the correct one.
So for concurrent SSTs, donors were swapping names.

Now, it is handled by a map, indexed by a donor name. To know if a node
is actual donor or not, it now compare timestamps of events. It assumes
both "selected donor" and "shifting DONOR" messages should have happen
in less than 0.01 secs to avoid any conflict.

Regression tests coming in next commit with an operator logs having
concurrent SSTs. Another conflicts was sometimes breaking the test
depending on the order on which we read files, hence why it's not added
here yet

ylacancellera committed:
Remove: propagation of ip propagation to older hash, Add: operator regression test

When adding "ownip", it was also propagating the new IP to the old hash.
But with operators, when IP where changing hash will also change, so
linking the new IP to old hash is anachronic. It is not wrong, but
depending on the order of map merges, the newest information could have
been overriden depending on the order of events.

That situation was producing X(2*number of conflicts) versions of output for operators, with
different md5sum, which could produce false-positive regression tests

So currently some information are not linked anymore, so some IPs are
not translated even though they could, but it is a limitation of using
maps as source of truth, as they are not versioned

ylacancellera committed:
Improve: main_test

It must test multiple times to remove doubts.
As the tool is reading files and relying on maps, their access order are
random. It can impact some translations

ylacancellera committed:
Add: operator member assocations regex

It existed for non-operator setup, but was not working for operators due
to k8s logs not interpreting newlines and tabs
This operator version re-uses existing regular regex handlers directly

ylacancellera commented:
> ./pt-galera-log-explainer list --all --since=2023-05-28T21:18:22.070376-05:00 --until=2023-05-29T08:19:02.849987-05:00 $HOME/test_pt/*20230528.log -vvv so verbosity options like -v, -vv, -vvv should be improved

Improved. Only -v and -vv exists now.
-v: display "useless" MySQL logs, only useful for the tool to translate stuff, eg "0509936f-8cc2 is node3"
-vv: script debug logs

ylacancellera commented:
> It is good to have support for test environments with the same hostname and without wsrep node name.

Will not support. I have added it to the limitations. Sharing IP and/or node names breaks identification, leading to errors.
Some level of ambiguity can be resolved (same ip having 2 current names because a garb was launched), but having a whole cluster on localhost with same name and hostname breaks every identification 

In my opinion, it's very okay  to not support it as a galera cluster on a single server is a terrible idea for production anyway.

ylacancellera commented:
TLDR: all new commits are to improve accuracy or fixing bugs, not adding new features.

Current state:
- some small corner-cases found thanks to YunusU (some people set the node public ip as wsrep_node_name apparently)
- regressions tests on upgrades, scattered logs, operators
- concurrent SSTs added (operators launching 2 garbds at the same time). It was not distinguished correctly. Regression tests for this added
- fixed some ambiguities of translations on operators, because of shared resources

Will try to refactor/simplify translations into a shared struct instead of having maps into each contexts.
Why: current implementation will overwrite translations where ip and names are re-used and exchanged between nodes (k8s).
Having timestamps for each info will make the whole translations more stable and easier to master.
This should simplify lots of code, if not it won't be committed.

ylacancellera committed:
Add: same ip/name limitation on README.rst

ylacancellera committed:
Add: shortuuid check, new date layout found

ylacancellera committed:
Add: inconsistent vote regex corner-case

ylacancellera committed:
Refactoring: migrate translations to singleton

It was using maps in each context, which would be merged between
contexts, then injected each time we needed a message to display.

It had a limitation on complicated operator setups: historical
information would be overriden by newer associations.
(e.g, that IP was for node0 yesterday, now it's node1, so associations
have been overwritten and incorrect)

It also introduced complexity, such as forcing to define closures too
many times, merging maps, it would be harder to debug, and every files
were starting from empty translation maps.
Moreover, iterating on maps is guaranteed to be random so it could create
hard-to-debug output variations on complex cases.

Now it is a singleton in translate package, still using maps but now it
associates an array of "units" storing the timestamp with each piece of information.
It is protected by rwmutex, because map are not threadsafe. (there's no
parallel processing for now)

No regressions, and it passes "operator_ambiguous_ips_list_all_no_color"
where the old system failed.
It nows also can be used as an easy to read source of information in
itself

ylacancellera commented:
Current state:
Refactored (and simplified, imo) the translation maps into a singleton.
It is now more complete, and completely stable so far to the worst operator logs I had yet.

Just need to re-adapt "whois" command, but it's a minor functionality.

ylacancellera committed:
Commenting out whois+sed, will be re-added later

ylacancellera committed:
Add: translate tests

ylacancellera commented:
Rebased on 3.x
"whois" and "sed" command are commented out: after the refactor, they seems complicated to get right and I'm not sure about the end-user formatting I should use, it has consequence on implementation. So better to keep it on a next PR, it's not the major functionality.

Ready to review, with regression tests, unit tests on every valuable functions (as far as I know, 90%+ of coverage on regexes and associated handlers)

ylacancellera committed:
Remove old comments, dead code

svetasmirnova commented:
Verbosity -v and -vvv seems to be the same:

```
sveta@s76:~/src/issues/CS0036005$ /home/sveta/src/percona/percona-toolkit/bin/pt-galera-log-explainer list --sst -v mysql-errordb2.log.10K
identifier                    Node2                                
current path                  mysql-errordb2.log.10K               
last known ip                 192.168.212.184                      
last known name               Node2                                
mysql version                 8.0.28                               
                                                                   
2023-09-12T20:47:34.249761Z   received shutdown                    
2023-09-12T20:47:44.271098Z   my_idx=0                             
2023-09-12T20:47:44.271240Z   SYNCED -> OPEN                       
2023-09-12T20:47:44.271316Z   OPEN -> CLOSED                       
                              my_idx=0                             
                              view member count: 1                 
                              8257a3a8-b0e2 is Node2               
                              my_idx=-1                            
                              view member count: 0                 
2023-09-12T20:49:04.555796Z   shutdown complete                    
2023-09-12T22:32:11.623838Z   starting(8.0.28)                     
2023-09-12T22:32:11.649205Z   started(cluster)                     
2023-09-12T22:39:49.819185Z   192.168.212.184 is local             
2023-09-12T22:39:49.963901Z   192.168.212.184 is local             
2023-09-12T22:39:49.966804Z   485a6520-a9ed is local               
2023-09-12T22:39:49.966930Z   485a6520-a9ed is local               
2023-09-12T22:39:49.970031Z   485a6520-a9ed is local               
2023-09-12T22:39:51.464022Z   CLOSED -> OPEN                       
2023-09-12T22:39:51.464526Z   my_idx=0                             
2023-09-12T22:39:51.466988Z   OPEN -> PRIMARY                      
2023-09-12T22:39:51.467336Z   485a6520-a9ed is local               
2023-09-12T22:39:52.360530Z   will receive IST(seqno:5084031493)   
2023-09-12T22:39:52.363278Z   Node1 will resync local node         
2023-09-12T22:39:52.363316Z   PRIMARY -> JOINER                    
2023-09-12T22:39:53.463996Z   485a6520-a9ed is local               
2023-09-12T22:39:56.034943Z   IST running                          
2023-09-12T22:39:56.884673Z   got IST from Node1                   
2023-09-12T22:40:43.835472Z   wsrep recovery                       
                              my_idx=-1                            
                              view member count: 3                 
                              80e7f955-8dfc is Node1               
                              8257a3a8-b0e2 is Node2               
                              c31e7e42-90c7 is garb                
                              my_idx=-1                            
                              view member count: 2                 
                              80e7f955-8dfc is Node1               
                              c31e7e42-90c7 is garb                
                              my_idx=0                             
                              view member count: 3                 
                              485a6520-a9ed is Node2               
                              80e7f955-8dfc is Node1               
                              c31e7e42-90c7 is garb                
2023-09-12T22:59:55.018362Z   IST received(seqno:5084031493)       
2023-09-12T22:59:55.021007Z   JOINER -> JOINED                     
2023-09-12T23:03:02.361034Z   JOINED -> SYNCED                     
sveta@s76:~/src/issues/CS0036005$ /home/sveta/src/percona/percona-toolkit/bin/pt-galera-log-explainer list --sst -vvv mysql-errordb2.log.10K
identifier                    Node2                                
current path                  mysql-errordb2.log.10K               
last known ip                 192.168.212.184                      
last known name               Node2                                
mysql version                 8.0.28                               
                                                                   
2023-09-12T20:47:34.249761Z   received shutdown                    
2023-09-12T20:47:44.271098Z   my_idx=0                             
2023-09-12T20:47:44.271240Z   SYNCED -> OPEN                       
2023-09-12T20:47:44.271316Z   OPEN -> CLOSED                       
                              my_idx=0                             
                              view member count: 1                 
                              8257a3a8-b0e2 is Node2               
                              my_idx=-1                            
                              view member count: 0                 
2023-09-12T20:49:04.555796Z   shutdown complete                    
2023-09-12T22:32:11.623838Z   starting(8.0.28)                     
2023-09-12T22:32:11.649205Z   started(cluster)                     
2023-09-12T22:39:49.819185Z   192.168.212.184 is local             
2023-09-12T22:39:49.963901Z   192.168.212.184 is local             
2023-09-12T22:39:49.966804Z   485a6520-a9ed is local               
2023-09-12T22:39:49.966930Z   485a6520-a9ed is local               
2023-09-12T22:39:49.970031Z   485a6520-a9ed is local               
2023-09-12T22:39:51.464022Z   CLOSED -> OPEN                       
2023-09-12T22:39:51.464526Z   my_idx=0                             
2023-09-12T22:39:51.466988Z   OPEN -> PRIMARY                      
2023-09-12T22:39:51.467336Z   485a6520-a9ed is local               
2023-09-12T22:39:52.360530Z   will receive IST(seqno:5084031493)   
2023-09-12T22:39:52.363278Z   Node1 will resync local node         
2023-09-12T22:39:52.363316Z   PRIMARY -> JOINER                    
2023-09-12T22:39:53.463996Z   485a6520-a9ed is local               
2023-09-12T22:39:56.034943Z   IST running                          
2023-09-12T22:39:56.884673Z   got IST from Node1                   
2023-09-12T22:40:43.835472Z   wsrep recovery                       
                              my_idx=-1                            
                              view member count: 3                 
                              80e7f955-8dfc is Node1               
                              8257a3a8-b0e2 is Node2               
                              c31e7e42-90c7 is garb                
                              my_idx=-1                            
                              view member count: 2                 
                              80e7f955-8dfc is Node1               
                              c31e7e42-90c7 is garb                
                              my_idx=0                             
                              view member count: 3                 
                              485a6520-a9ed is Node2               
                              80e7f955-8dfc is Node1               
                              c31e7e42-90c7 is garb                
2023-09-12T22:59:55.018362Z   IST received(seqno:5084031493)       
2023-09-12T22:59:55.021007Z   JOINER -> JOINED                     
2023-09-12T23:03:02.361034Z   JOINED -> SYNCED   
```
Please adjust README file.

svetasmirnova commented on src/go/pt-galera-log-explainer/README.rst:
```diff
@@ -0,0 +1,240 @@
+    Remove regexes from analysis. Use 'pt-galera-log-explainer regex-list' to have the list
+    
+``--grep-cmd``
+    grep v3 binary command path. For Darwin systems, it could need to be set to 'ggrep'
```
```suggestion
    grep v3 binary command path. For Darwin systems, it could need to be set to ``ggrep``
```

svetasmirnova commented on src/go/pt-galera-log-explainer/README.rst:
```diff
@@ -0,0 +1,240 @@
+    Default: ``grep``
+
+``--grep-args``     
+    grep arguments. perl regexp (-P) is necessary. -o will break the tool
```
This quite not clear. Please explain and/or add examples.

ylacancellera replied:
This argument is not helpful in the end, removing entirely

svetasmirnova reviewed (changes requested):
Please check comments and adjust accordignly.

svetasmirnova commented:
Please also fix typos (see spelling/Typos report). You can leave false-positives (like IST) for me or fix .typos.toml as explained in https://github.com/crate-ci/typos/blob/c97d621b6b01d8b0258538ca15abeca5c5764601/README.md and https://github.com/crate-ci/typos/tree/c97d621b6b01d8b0258538ca15abeca5c5764601/docs

artemgavrilov commented on src/go/pt-galera-log-explainer/conflicts.go:
```diff
@@ -0,0 +1,72 @@
+		}
+		var out string
+
+		if c.Yaml {
```
Instead of if-else we can use here switch without condition: https://go.dev/tour/flowcontrol/11
```
switch {
  case c.Yaml:
    // process yaml
  case c.Json: 
    // process json
  defalut:
    // do the rest
}
```

ylacancellera replied:
Agreed

artemgavrilov commented on src/go/pt-galera-log-explainer/conflicts.go:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	ctxs := timeline.GetLatestContextsByNodes()
+	for _, ctx := range ctxs {
```
Minor: `ctx` and `context` has some common meaning in golang. Probably it's not the best name for that variable and entity.

ylacancellera replied:
@artemgavrilov agree. Though, in a code with no mention of HTTP, I assumed it could be ok
Do you have a suggestion of a better-fitting name ?

artemgavrilov commented on src/go/pt-galera-log-explainer/conflicts.go:
```diff
@@ -0,0 +1,72 @@
+		} else {
+
+			for _, conflict := range ctx.Conflicts {
+				out += "\n"
+				out += "\n" + utils.Paint(utils.BlueText, "seqno: ") + conflict.Seqno
+				out += "\n\t" + utils.Paint(utils.BlueText, "winner: ") + conflict.Winner
+				out += "\n\t" + utils.Paint(utils.BlueText, "votes per nodes:")
```
Minor: strings builder can be more optimal solution here: https://pkg.go.dev/strings#Builder

ylacancellera replied:
Agreed for this function, the results can actually be pretty long so I'm applying it

artemgavrilov commented on src/go/pt-galera-log-explainer/conflicts.go:
```diff
@@ -0,0 +1,72 @@
+
+		}
+		fmt.Println(out)
+		return nil
```
It's redundant
```suggestion
```

ylacancellera replied:
This one should be left it as it is for now
It's not actually redundant, it's to make sure we only print 1 list of conflicts, and not one per file given in parameter
The real fix would have been to complete those "conflicts" between every nodes

ylacancellera commented on src/go/pt-galera-log-explainer/internal.go:
```diff
@@ -0,0 +1,218 @@
+		// We have to find again what regex worked to get this log line
+		// it can match multiple regexes
+		for key, regex := range regexes {
+			if !regex.Regex.MatchString(line) || utils.SliceContains(CLI.ExcludeRegexes, key) {
```
fyi, it's by far the current bottleneck (around 35%-40% of total CPU time spent on medium files in regression tests with pprof)
My idea: make a quick estimation using "strings.Contains(line, word)" for every word in the regexes. So that it's strict comparison, and not regex anymore which is too slow
Few false positives would be okay anyway, as regexes are re-checked with "internalRegex" in the next line in "regex.Handle"

artemgavrilov commented on src/go/pt-galera-log-explainer/ctx.go:
```diff
@@ -0,0 +1,42 @@
+	"github.com/percona/percona-toolkit/src/go/pt-galera-log-explainer/translate"
+)
+
+type ctx struct {
```
Really not the best name for the structure.

ylacancellera replied:
@artemgavrilov In this context this defines the [kong](https://github.com/alecthomas/kong) command name
It was making sense when using the CLI
Got a better suggestion ?

ylacancellera replied:
Will only keep this ctx, as it's about CLI usage. I removed the others ctx though

artemgavrilov commented on src/go/pt-galera-log-explainer/display/timelinecli.go:
```diff
@@ -0,0 +1,352 @@
+	// a slice keeps its order
+	keys, currentContext := initKeysContext(timeline)    // currentcontext to follow when important thing changed
+	latestContext := timeline.GetLatestContextsByNodes() // so that we have fully updated context when we print
+	lastContext := map[string]types.LogCtx{}             // just to follow when important thing changed
```
Minor: In PMM we are following this advise: https://dave.cheney.net/practical-go/presentations/qcon-china.html#_use_a_consistent_declaration_style


So I suggest to create empty maps like this:
```suggestion
	lastContext := make(map[string]types.LogCtx)              // just to follow when important thing changed
```

ylacancellera replied:
Agree, because actually I could prealloc the correct size here

ylacancellera committed:
Fix: typos

ylacancellera commented:
Some typos will have to stay, they are from Galera lib itself:

scripts/wsrep_sst_xtrabackup-v2.sh:1380:        wsrep_log_error "Possible timeout in **receving** first data from donor in gtid/keyring stage"
percona-xtradb-cluster-galera/galera/src/saved_state.cpp:54:        log_warn << "No persistent state found. **Bootstraping** with default state";

ylacancellera committed:
Remove: --grep-args

Can be used to break the tool, and I actually never had an use-case

ylacancellera committed:
PR-669 - Add pt-galera-log-explainer percona#669

- Commented out whois command in README.rst
- Copied README.rst into docs directory, so user manual will include
  this tool

ylacancellera committed:
Fix: README with -vvv, --grep-args, regex-list

ylacancellera committed:
Update src/go/pt-galera-log-explainer/README.rst

Co-authored-by: Sveta Smirnova <svetasmirnova@users.noreply.github.com>

ylacancellera commented:
My mistake on a git local state, I remade the commit on docs/pt-galera-log-explainer.rst

Apart from that, changes according to comments have been applied

artemgavrilov commented on src/go/pt-galera-log-explainer/display/timelinecli.go:
```diff
@@ -0,0 +1,352 @@
+	return out
+}
+
+func (ts *transitions) iterate() {
```
It looks like this method counts something, probably is should have a different name.

ylacancellera replied:
"iterate" was named that way to help readability in the main "transitionSeparator" func
I would vote to keep it that way

artemgavrilov commented on src/go/pt-galera-log-explainer/display/timelinecli_test.go:
```diff
@@ -0,0 +1,186 @@
+	"github.com/percona/percona-toolkit/src/go/pt-galera-log-explainer/utils"
+)
+
+func TestTransitionSeparator(t *testing.T) {
+	tests := []struct {
```
It seems it's safe to run this test in parallel:
```suggestion
func TestTransitionSeparator(t *testing.T) {
        t.Parallel()
	tests := []struct {
```

ylacancellera replied:
Agreed, committed

artemgavrilov commented on src/go/pt-galera-log-explainer/internal.go:
```diff
@@ -0,0 +1,218 @@
+
+func execGrepAndIterate(path, compiledRegex string, stdout chan<- string) error {
+
+	defer close(stdout)
```
I suggest to close this channel outside this function.

ylacancellera replied:
Done, it does make more sense

artemgavrilov commented on src/go/pt-galera-log-explainer/internal.go:
```diff
@@ -0,0 +1,218 @@
+
+	cmd := exec.Command(CLI.GrepCmd, "-P", compiledRegex, path)
+
+	out, _ := cmd.StdoutPipe()
```
Better to not ignore errors, it can save a lot of time later.

ylacancellera replied:
Agreed, my mistake

artemgavrilov commented on src/go/pt-galera-log-explainer/regex/file_test.go:
```diff
@@ -0,0 +1,47 @@
+	"testing"
+)
+
+func TestFileType(t *testing.T) {
```
```suggestion
func TestFileType(t *testing.T) {
    t.Parallel()
```

ylacancellera replied:
Done (in another commit, among others things)

artemgavrilov commented on src/go/pt-galera-log-explainer/sed.go:
```diff
@@ -0,0 +1,102 @@
+package main
+
+/*
+type sed struct {
+	Paths []string `arg:"" name:"paths" help:"paths of the log to use"`
```
Do we need this file?

ylacancellera replied:
Seen with Sveta, we can keep
It should have been included in this PR but it will re-added in another one, there's an usage debate with that one

artemgavrilov commented on src/go/pt-galera-log-explainer/translate/translate.go:
```diff
@@ -0,0 +1,206 @@
+
+	// incase methods changed in the middle, tls=>ssl
+	IPToMethods map[string][]translationUnit
+	rwlock      *sync.RWMutex
```
Usually there are no reasons to have pointer to mutex.
```suggestion
	rwlock      sync.RWMutex
```

artemgavrilov commented on src/go/pt-galera-log-explainer/translate/translate.go:
```diff
@@ -0,0 +1,206 @@
+		HashToNodeNames: map[string][]translationUnit{},
+		IPToMethods:     map[string][]translationUnit{},
+		IPToNodeNames:   map[string][]translationUnit{},
+		rwlock:          &sync.RWMutex{},
```
If to use mutex by value, then we don't need to initialize it at all.
```suggestion
```

ylacancellera replied:
Thanks, I was not sure

artemgavrilov commented on src/go/pt-galera-log-explainer/translate/translate.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+func DBToJson() string {
+	out, _ := json.MarshalIndent(db, "", "\t")
```
Unhandled error may cause troubles sometime.

ylacancellera replied:
Agree

artemgavrilov commented on src/go/pt-galera-log-explainer/types/conflicts.go:
```diff
@@ -0,0 +1,57 @@
+	Error string
+}
+
+func (cs Conflicts) Merge(c Conflict) Conflicts {
```
It looks that it makes sense to use pointer here and in all other methods.
```suggestion
func (cs *Conflicts) Merge(c Conflict) Conflicts {
```

ylacancellera replied:
@artemgavrilov "Conflicts" is a slice of pointers, Conflicts []*Conflict, because they can be shared
Which is why I kept a value ref, as pointers to slice would not serve much as far as I know
(it would also force an iteration over "range *cs")

artemgavrilov commented on src/go/pt-galera-log-explainer/types/sst.go:
```diff
@@ -0,0 +1,23 @@
+	SelectionTimestamp *time.Time
+}
+
+// MustHaveHappenedLocally use the "selected x as donor" timestamp
+// and compare it to the timestamp of donor/joiner wsrep status shift
+// Usually, when it is selected, joiner/donor take a few milliseconds to shift their status
+// This is the most solid way so far to correctly map donor and joiners when concurrents SSTs
+// are running
```
😲

ylacancellera replied:
I'm actually the first one disappointed. There's just not enough metadata printed to make sure.
I actually got cases with 2 SSTs starting at the same time (more or less), each with their donors and joiners, so when files don't have much info yet and we still haven't got each others names (which actually happen on k8s), that's the only reliable way I have left
The false negatives from this method were less impacting for readability, hence why I kept it

artemgavrilov commented on src/go/pt-galera-log-explainer/types/timeline.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+func (t *Timeline) GetLatestContextsByNodes() map[string]LogCtx {
+	latestctxs := map[string]LogCtx{}
```
Minor: can be preallocated with proper size

ylacancellera replied:
agree

artemgavrilov commented on src/go/pt-galera-log-explainer/whois.go:
```diff
@@ -0,0 +1,94 @@
+package main
+
+/*
+type whois struct {
+	Search string   `arg:"" name:"search" help:"the identifier (node name, ip, uuid, hash) to search"`
```
Also, do we need this file to be commited?

ylacancellera replied:
Seen with Sveta as well, can keep
Will be re-added later properly, the latest commits commented it out.

ylacancellera committed:
Add: parallel on 2 unit tests

egegunes commented on src/go/pt-galera-log-explainer/types/ctx.go:
```diff
@@ -0,0 +1,222 @@
+	if !utils.SliceContains([]string{"SYNCED", "JOINED", "DONOR", "DESYNCED", "JOINER", "PRIMARY", "NON-PRIMARY", "OPEN", "CLOSED", "DESTROYED", "ERROR", "RECOVERY"}, s) {
+		return
+	}
+	//ctx.state[ctx.FileType] = append(ctx.state[ctx.FileType], s)
```
nitpick: no need to commit commented code

egegunes commented on src/go/pt-galera-log-explainer/types/conflicts.go:
```diff
@@ -0,0 +1,57 @@
+		}
+	}
+	return nil
+}
```
nitpick: there should be a newline between functions

egegunes reviewed (approved):
I commented two nitpicks, they're up to you.

Apart from them, I see `ctx` is commonly used as variable name for type `LogCtx`. `ctx` is conventionally used for `context.Context` in go. Of course it's not written in stone but this may confuse other developers reading the code.

ylacancellera committed:
Imp: use strings.builder for conflicts

ylacancellera committed:
Fix: missed errors, minor formatting issues

ylacancellera committed:
Rename ctx to logCtx, remove any mention of ctx

ylacancellera commented:
Changed every "ctx" mention to logCtx

ylacancellera committed:
Move: chan closing in "main" func

ylacancellera committed:
preallocate lastcontexts map

ylacancellera commented:
@svetasmirnova review changes have been applied. I explained and resolved the few tiny changes I did not apply. I currently have nothing left to do for this PR for now

artemgavrilov reviewed (approved):
Sorry for the delay in re-review. LGTM.

svetasmirnova committed:
PR-669 -  Add pt-galera-log-explainer

- Sync README.rst and docs/pt-galera-log-explainer.rst

svetasmirnova committed:
Merge branch '3.x' into galera-log-explainer

svetasmirnova committed:
PR-669 - Add pt-galera-log-explainer

- Added exception for IST into .typos.toml

svetasmirnova committed:
PR-669 -  Add pt-galera-log-explainer

- Added exception into .typos.toml for src/go/pt-galera-log-explainer/tests/

svetasmirnova committed:
PR-669 -  Add pt-galera-log-explainer

- Added exception for Go tests into .typos.toml

svetasmirnova committed:
PR-669 -  Add pt-galera-log-explainer

- Exceptions for identifiers and Galera library errors in .typos.toml
